### PR TITLE
was14rdma

### PR DESCRIPTION
**Pull request by was4**
See extensive commit header for details. Authentication not tested, but mostly simple symbol changes by inspection. After discussion with Marcus, leaving more substantive GSS changes until another pass.

### CODE REVIEW

File: ntirpc/rpc/auth_inline.h [hunk 1]
```diff
@@ -156,10 +153,6 @@ inline_auth_decode(XDR *xdrs, struct opaque_auth *oa, int32_t *buf)
 static inline bool
 inline_auth_free(XDR *xdrs, struct opaque_auth *oa)
 {
-	if (oa->oa_base != NULL) {
```
Comment (mattbenjamin) [inline ntirpc/rpc/auth_inline.h:159]: @was4  cool

File: ntirpc/rpc/rpc_msg.h [hunk 2]
```diff
@@ -103,6 +102,9 @@ struct accepted_reply {
 	} ru;
 #define ar_results ru.AR_results
 #define ar_vers  ru.AR_versions
+
+	/* after union to avoid (rare) corruption by rejected_reply */
+	struct opaque_auth ar_verf;
```
Comment (mattbenjamin) [inline ntirpc/rpc/rpc_msg.h:107]: ok

File: ntirpc/rpc/rpc_msg.h [hunk 3]
```diff
@@ -139,11 +141,6 @@ struct reply_body {
  */
 struct call_body {
 	rpcvers_t cb_rpcvers;	/* must be equal to two */
-	rpcprog_t cb_prog;
```
Comment (mattbenjamin) [inline ntirpc/rpc/rpc_msg.h:142]: ok

File: src/svc.c [hunk 4]
```diff
@@ -230,39 +230,13 @@ svc_init(svc_init_params *params)
 	return true;
 }
 
-struct rpc_msg *
-alloc_rpc_msg(void)
-{
-	struct rpc_msg *msg = mem_alloc(sizeof(struct rpc_msg));
-
-	TAILQ_INIT_ENTRY(msg, msg_q);
```
Comment (mattbenjamin) [inline src/svc.c:238]: @was4 so where has this moved?  this is needed for bi-directional

**Review by mattbenjamin — COMMENTED** [review]
@was4 I strongly approve of removing the allocation, but at the same time I introduced it to cope with bi-directional calls where a message that svc_recv() decoded might need to be returned (and in the model as it was) freed by clnt_call().  I'm guessing that changes to clnt_call() are needed?

**Comment by mattbenjamin** [comment]
@was4 see above--I think you're on the right track, but I think that as is, this will break clnt_call w/bi-directional; presumably can be adapted--I agree, I'd rather not impose another allocation of rpc_msg going forward

**Comment by mattbenjamin** [comment]
updated

Matt

----- Original Message -----
> From: "was4" <notifications@github.com>
> To: "nfs-ganesha/ntirpc" <ntirpc@noreply.github.com>
> Cc: "Matt Benjamin" <mbenjamin@redhat.com>, "Mention" <mention@noreply.github.com>
> Sent: Friday, December 16, 2016 2:46:19 PM
> Subject: Re: [nfs-ganesha/ntirpc] was14rdma (#30)
> 
> @was4 requested your review on: nfs-ganesha/ntirpc#30 was14rdma.
> 
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/nfs-ganesha/ntirpc/pull/30#event-897162402

-- 
Matt Benjamin
Red Hat, Inc.
315 West Huron Street, Suite 140A
Ann Arbor, Michigan 48103

http://www.redhat.com/en/technologies/storage

tel.  734-821-5101
fax.  734-769-8938
cel.  734-216-5309

**Review by dang — APPROVED** [review]
Seems fine to me, but I'm not an expert in this code.

File: src/svc.c (hunk 4 above)
Comment (was4) [reply src/svc.c:238]: It hasn't moved. It is never used anywhere. I'm trying to figure out what you meant, but see nothing in the rpc_call_ctx other than the rb tree.

In the ganesha nfs_core.h, there is a struct glist_head req_q; is that what you mean?

Comment (mattbenjamin) [reply src/svc.c:238]: I'm worried about the rpc_msg itself.  In client call, iirc, we allocate one, but when bi-directional, we get a pointer to one allocated in svc_recv (or its caller?) and work with it.

Comment (was4) [reply src/svc.c:238]: Ah. Now copied. No alloc/free.

> 		ctx->cc_msg = *msg;	/* and stash reply header */

Comment (mattbenjamin) [reply src/svc.c:238]: yeah, ok, that's what I was looking for

**Comment by was4** [comment]
Pushed a new version with the conflict (of an hour ago) resolved.

**Pull request merged**